sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/uart.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/log.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_plic.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/check.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/testing/test_status.h"
+
+#define UART_DATASET_SIZE 128
+#define UART0_BASE_ADDR 0x40000000u
+#define PLIC0_BASE_ADDR 0x40090000u
+
+static dif_uart_t uart0;
+static dif_plic_t plic0;
+
+/**
+ * UART TX RX test
+ *
+ * This test sends and receives a known dataset over UART. The size of the
+ * dataset is indicated with UART_DATASET_SIZE. The dataset is agreed upon by
+ * the device (a.k.a. the OpenTitan chip) and the host (a.k.a. the simulation
+ * device, such as DV testbench). communicating with it. Data transmitted over
+ * TX is checked for correctness at the host, and likewise, data sent by the
+ * host is checked for correctness at the device (in this SW test). The data
+ * transmitted over TX and RX ports may occur simultaneously. The test ensures
+ * that the TX watermark, RX watermark and TX empty interrupts are seen.
+ * At the end, the host transmits another set of random data (greater than the
+ * RX fifo size which the device drops, to generate the RX overflow condition.
+ * The test passes when the datasets at both ends match the expected and all
+ * required intererupts are seen.
+ */
+
+// TODO: Make these datasets random.
+
+// A set of bytes to be send out of TX.
+static const uint8_t uart_tx_data[UART_DATASET_SIZE] = {
+    0xe8, 0x50, 0xc6, 0xb4, 0xbe, 0x16, 0xed, 0x55, 0x16, 0x1d, 0xe6, 0x1c,
+    0xde, 0x9f, 0xfd, 0x24, 0x89, 0x81, 0x4d, 0x0d, 0x1a, 0x12, 0x4f, 0x57,
+    0xea, 0xd6, 0x6f, 0xc0, 0x7d, 0x46, 0xe7, 0x37, 0x81, 0xd3, 0x8e, 0x16,
+    0xad, 0x7b, 0xd0, 0xe2, 0x4f, 0xff, 0x39, 0xe6, 0x71, 0x3c, 0x82, 0x04,
+    0xec, 0x3a, 0x27, 0xcc, 0x3d, 0x58, 0x0e, 0x56, 0xd2, 0xd2, 0xb9, 0xa3,
+    0xb5, 0x3d, 0xc0, 0x40, 0xba, 0x90, 0x16, 0xd8, 0xe3, 0xa4, 0x22, 0x74,
+    0x80, 0xcb, 0x7b, 0xde, 0xd7, 0x3f, 0x4d, 0x93, 0x4d, 0x59, 0x79, 0x88,
+    0x24, 0xe7, 0x68, 0x8b, 0x7a, 0x78, 0xb7, 0x07, 0x09, 0x26, 0xcf, 0x6b,
+    0x52, 0xd9, 0x4c, 0xd3, 0x33, 0xdf, 0x2e, 0x0d, 0x3b, 0xab, 0x45, 0x85,
+    0xc2, 0xc2, 0x19, 0xe5, 0xc7, 0x2b, 0xb0, 0xf6, 0xcb, 0x06, 0xf6, 0xe2,
+    0xf5, 0xb1, 0xab, 0xef, 0x6f, 0xd8, 0x23, 0xfd,
+};
+static size_t uart_tx_bytes_written;
+
+// A set of bytes actually received over RX.
+static uint8_t uart_rx_data[UART_DATASET_SIZE];
```

Comment:
This should probably be passed in as a pointer to the receive function, and allocated on `main`'s stack, instead of being a global.